proof/bls_operations.saw

```diff
@@ -0,0 +1,1093 @@
+                        (e1_scalar_mult (scalar_value`{255,32} sk)
+                                        (affinify E (hash_to_curve_opt_impl x)))) &&
+                  ~(is_bad_O_form_E P)) }});
+
+let remove_ho_rules = concat
+  [ point_add_fp2_thm, point_dadd_fp2_thm, point_double_fp2_thm, point_neg_fp2_thm
+   , point_add_affine_fp2_thm, point_dadd_affine_fp2_thm
+   , add_E'_thm, add'_E'_thm, neg_E'_thm
+   , affinify_E'_thm, projectify_E'_thm
+   , is_point_affine_E'_thm, is_point_projective_E'_thm
+   , is_point_O_E'_thm ]
+  [ point_add_fp_thm, point_dadd_fp_thm, point_double_fp_thm, point_neg_fp_thm
+   , point_add_affine_fp_thm, point_dadd_affine_fp_thm
+   , add_E_thm, add'_E_thm, affinify_E_thm,  projectify_E_thm
+   , is_point_affine_E_thm, is_point_projective_E_thm, is_point_O_E_thm, point_O_E_thm
+   ];
+
+
+test_thm_correct' <- custom_prove_cryptol (rewrite (cryptol_ss())
+  {{ \P sk (x:([32][8],[43][8])) ->
+                (same_point_affine
+                  E
+                  (affinify E (POINTonE1_abs P))
+                  (e1_scalar_mult (scalar_value`{255,32} sk)
+                                  (affinify E (hash_to_curve_opt_impl x)))) ==
+                ((apply same_point_affine
+                        E
+                        (affinify E (POINTonE1_abs P))
+                        (e1_scalar_mult (scalar_value`{255,32} sk)
+                                        (affinify E (hash_to_curve_opt_impl x)))) &&
+                  POINTonE1_invariant P &&
+                  ~(is_bad_O_form_E P) &&
+                  is_point_projective E (POINTonE1_abs P)) }})
+  do {
+    rw_with_1 e1_scalar_mult_not_bad_O_thm;
+    unfolding ["apply"];
+    simplify basic_ss;
+    simplify (cryptol_ss());
+    rw_with POINTonE1_base_thms;
+    simplify (addsimps [ affine_inv_thm
+                       , e1_scalar_mult_on_curve_thm
+                       , hash_to_g1_on_curve_thm
+                       , affine_inv_affine_rep_e1_thm ]
+                       empty_ss);
+    simplify basic_ss;
+    rw_with remove_ho_rules;
+    rw_with_1 same_point_affine_E_thm;
+    rw_with_1 and_true_thm;
+    print_goal;
+    w4_unint_z3 (concat [ "same_point_affine_E"
+                        , "hash_to_curve_opt_impl"
+                        , "scalar_value"
+                        , "is_bad_O_form_E" ]
+                        ec_mult_unints);
+  };
+
+hash_to_curve_e1_opt_thm <- admit_cryptol
+  {{ \x -> HE1::hash_to_curve x == HE1::hash_to_curve_opt x }};
+
+hash_to_curve_arg_expand_thm <- prove_cryptol
+  {{ \(x:([32][8],[43][8])) ->
+       hash_to_curve_opt_impl x ==
+       apply hash_to_curve_opt_impl (([]:[0][8])#x.0, x.1) }} ["hash_to_curve_opt_impl"];
+
+// TODO: why is this needed here?
+hash_to_curve_E1_opt_impl_equiv_ov <- (prove_hash_to_curve_E1_opt_impl_equiv_thm 32 43);
+
+// TODO: why is this needed here (see prove_hash_to_g1_ov in hash_to_g1.saw)?
+hash_to_curve_e1_impl_thm <- custom_prove_cryptol
```

Comment:
Could we use `blst_hash_to_g1_spec` instead so that we get the high-level spec directly?

Reply:
Maybe?  The reason I used `blst_hash_to_g1_impl_spec` is that it's postcondition is written with `llvm_points_to`, while `blst_hash_to_g1_spec` is written with `llvm_postcond`.  The `llvm_postcond` forms tend to be painful to work with.